src/storage/src/storage/upload_object/unbuffered.rs

```diff
@@ -35,6 +36,114 @@ where
     /// # Ok(()) }
     /// ```
     pub async fn send_unbuffered(self) -> Result<Object> {
+        let hint = self
+            .payload
+            .lock()
+            .await
+            .size_hint()
+            .await
+            .map_err(Error::deser)?;
+        let threshold = self.options.resumable_upload_threshold as u64;
+        if hint.1.is_none_or(|max| max >= threshold) {
+            self.send_unbuffered_resumable(hint).await
+        } else {
+            self.send_unbuffered_single_shot().await
+        }
+    }
+
+    async fn send_unbuffered_resumable(self, hint: (u64, Option<u64>)) -> Result<Object> {
+        let mut upload_url = None;
+        gax::retry_loop_internal::retry_loop(
+            async |_| self.resumable_attempt(&mut upload_url, hint).await,
+            async |duration| tokio::time::sleep(duration).await,
+            true,
+            self.options.retry_throttler.clone(),
+            Arc::new(ContinueOn308::new(self.options.retry_policy.clone())),
+            self.options.backoff_policy.clone(),
+        )
+        .await
+    }
+
+    async fn resumable_attempt(
+        &self,
+        url: &mut Option<String>,
+        hint: (u64, Option<u64>),
+    ) -> Result<Object> {
+        let offset = if let Some(upload_url) = url {
+            match self.query_resumable_upload(upload_url).await? {
+                ResumableUploadStatus::Finalized(o) => {
+                    return Ok(*o);
+                }
+                ResumableUploadStatus::Partial(o) => o,
+            }
+        } else {
+            let upload_url = self.start_resumable_upload_attempt().await?;
+            *url = Some(upload_url);
+            0_u64
+        };
+        assert!(url.is_some());
+
+        use crate::upload_source::Seek;
+        let payload = self.payload.clone();
+        payload
+            .lock()
+            .await
+            .seek(offset)
+            .await
+            .map_err(Error::ser)?;
+
+        let range = match (offset, hint.0, hint.1) {
+            (o, _, None) => format!("bytes {o}-*/*"),
+            (_, 0, Some(0)) => "bytes */0".to_string(),
+            (o, s, Some(u)) if s == u => format!("bytes {o}-{}/{s}", s - 1),
```

Comment:
I see that this is the case where we know the size of the object.

Q: So we do not have to break the payload into partial uploads and there is only one `Content-Range` header per stream?

https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
https://cloud.google.com/storage/docs/streaming-uploads#json-api

Reply:
> Q: So we do not have to break the payload into partial uploads and there is only one Content-Range header per stream?

No. That is the point of using unbuffered uploads with `Seek`: send all the data in a single PUT request. If it succeeds you get really good performance (all the TCP windows get large, the server can optimize writes, etc.).  If the PUT request fails, you query the status of the upload and send another PUT request from the last persisted byte: that is where having a `Seek` implementation helps.

Reply:
Thanks for the explainer.